src/lib/schematics/update/rules/ripple-v7/switchRippleSpeedFactorRule.ts

```diff
@@ -0,0 +1,114 @@
+import {ProgramAwareRuleWalker, RuleFailure, Rules} from 'tslint';
+import * as ts from 'typescript';
+import {bold, red, green} from 'chalk';
+import {convertSpeedFactorToDuration} from './ripple-speed-factor';
+
+/**
+ * Rule that walks through property assignment and switches the global `baseSpeedFactor`
+ * ripple option to the new global animation config.
+ * Also updates every class member assignment that refers to MatRipple#speedFactor.
+ */
+export class Rule extends Rules.TypedRule {
+  applyWithProgram(sourceFile: ts.SourceFile, program: ts.Program): RuleFailure[] {
+    return this.applyWithWalker(
+        new SwitchRippleSpeedFactorRule(sourceFile, this.getOptions(), program));
+  }
+}
+
+export class SwitchRippleSpeedFactorRule extends ProgramAwareRuleWalker {
+
+  /**
+   * Switches binary expressions (e.g. myRipple.speedFactor = 0.5) to the new animation config.
+   */
+  visitBinaryExpression(expression: ts.BinaryExpression) {
+    if (expression.left.kind !== ts.SyntaxKind.PropertyAccessExpression) {
+      return;
+    }
+
+    // Left side expression consists of target object and property name (e.g. myInstance.val)
+    const leftExpression = expression.left as ts.PropertyAccessExpression;
+    const targetTypeNode = this.getTypeChecker().getTypeAtLocation(leftExpression.expression);
+
+    if (!targetTypeNode.symbol) {
+      return;
+    }
+
+    const targetTypeName = targetTypeNode.symbol.getName();
+    const propertyName = leftExpression.name.getText();
+
+    if (targetTypeName === 'MatRipple' && propertyName === 'speedFactor') {
+      if (expression.right.kind === ts.SyntaxKind.NumericLiteral) {
+        const numericValue = parseFloat((expression.right as ts.NumericLiteral).text);
+        const newEnterDurationValue = convertSpeedFactorToDuration(numericValue);
+
+        // Replace the `speedFactor` property name with `animation`.
+        const propertyNameReplacement = this.createReplacement(leftExpression.name.getStart(),
+          leftExpression.name.getWidth(), 'animation');
+
+        const rightExpressionReplacement = this.createReplacement(expression.right.getStart(),
+          expression.right.getWidth(), `{enterDuration: ${newEnterDurationValue}}`);
+
+        this.addFailureAtNode(
+          expression,
+          `Found deprecated member assignment for "${bold('MatRipple')}#${red('speedFactor')}"`,
+          [propertyNameReplacement, rightExpressionReplacement]);
+      } else {
+        // In case the speed factor is dynamically calculated or passed to the assignment, we just
+        // print the failure and notify about the breaking change.
```

Comment:
Could you just treat the entire expression as the "value"?

Reply:
Not sure what you mean with that? Do you say we should just create the actual calculation instead of executing it?

```ts
r.speedFactor = mySpeedFactor

// --->

r.animation = {enterDuration: (450 / mySpeedFactor)};
```

Reply:
That's pretty much what I was thinking

Reply:
We can totally do that, but not sure how much that would help with upgrading. That's more like a workaround to just make the upgrade succeed.

My goal with the rules would be to generate code that can actually stay in long-term, and in that specific situation, people should switch their _whole_ (not just the assignment) code-base from speed factor to the new animation config

Reply:
That makes sense. I will make the change but cannot guarantee about the `TODO` because it's not clear how the actual assignment is placed in the file. If that won't work, I will either print a message, or just do an inline TODO.

e.g. 

`r.animation = /** TODO: Remove speed factor */ {enterDuration: (450 / mySpeedFactor)};`